plugins/catalog-backend/src/next/DefaultCatalogProcessingOrchestrator.ts

```diff
@@ -276,23 +293,46 @@ function createEmitter(logger: Logger, parentEntity: Entity) {
       );
       return;
     }
-    if (i.type === 'entity') {
-      // TODO(freben): Perform the most basic validation here
-      // (apiVersion, kind, metadata, metadata.name, metadata.namespace, spec)
 
-      const originLocation = getEntityOriginLocationRef(parentEntity);
+    if (i.type === 'entity') {
+      let entity: EntityEnvelope;
+      try {
+        entity = validateEnvelope(i.entity);
+      } catch (e) {
+        logger.debug(`Envelope validation failed at ${i.location}, ${e}`);
+        errors.push(e);
+        return;
+      }
 
-      deferredEntities.push({
-        ...i.entity,
-        metadata: {
-          ...i.entity.metadata,
-          annotations: {
-            ...i.entity.metadata.annotations,
-            [ORIGIN_LOCATION_ANNOTATION]: originLocation,
-            [LOCATION_ANNOTATION]: stringifyLocationReference(i.location),
+      // TODO(freben): Should annotations be part of the envelope after all?
+      // At least with very very basic validation (that if it exists, it's an
+      // object)? That would let us get rid of this type trickery. But at the
+      // same time, we want to defer "hard" validation to as late as possible
+      // so that any errors/warnings can end up in the context of the output
+      // entity rather than the location.
+      const annotations =
+        ((entity.metadata as any).annotations as Record<string, any>) || {};
+      if (
+        annotations &&
+        typeof annotations === 'object' &&
+        !Array.isArray(annotations)
```

Comment:
this code is intentionally defensive; if there was for example an annotations array instead of object, we want to just leave it alone for now, let it land in the processing state, and fail actual full validation later